yelp_scrape.py

```diff
@@ -23,16 +24,30 @@ def url_generator(search_item, location):
     search_item = search_item.replace(' ', '+')
     location = location.replace(' ', '+').replace(',', '%2C')
     url = base_url + 'find_desc=' + search_item + '&find_loc=' + location
-    return url
-
-
-search, loc = get_search_request()
-yelp_page = requests.get(url_generator(search, loc))
-yelp_soup = BeautifulSoup(yelp_page.content, 'html.parser')
-
-test = yelp_soup.find_all('a')
-links = []
-for a in test:
-    links.append(a["href"])
-
-print(links)
+    file_name = search_item + location
```

Comment:
This file name is likely not going to be valid because it'll contain all kinds of spaces and `.` and things.

Filesystems are extremely finnicky, and you don't wanna use characters in a filename that you're not 100% sure are supported by the operating system.

So here, I'd do something like
```python
hash_value: int = hash(search_item + location)
hash_str: str = hex(hash_value)
file_name: str = f"{hash_str}.html"
return url, file_name
```
the `hex()` function turns an integer into a base-16 encoded string (meaning it uses only 0-9 and a-f) so it's guaranteed to be a safe filename.

for hygiene and legibility, I'd also split out the hashing portion to a separate function so it's easy to reuse, and it's clear what it's for

something like `def hash_key(search_item: str, location: str) -> str`

Reply:
So when I create a test function like

def hash_test(test_string: str) ->str:
    hash_value: int = hash(test_string)
    hash_str: str = hex(hash_value)
    return hash_str

I get a different hash_str every time. Whenever I use a string as an input to hash(), it generates a different int on repeat runs despite using the same string for testing. This is that problem I was describing yesterday.

Reply:
hmmm... that certainly seems fucky

when something seems fucky, a good idea is to nail it down
so write a test function that fails because of this bug

something like 

```python
def test_hash_stuff():
  a = "same string"
  b = "same string"
  assert(hash_test(a) == hash_test(b))
```

when you run this, you _expect_ it to pass (ie, the assertion does not fail)
if it _does_ fail, it gives other people (and you) a small jumping off point to debug what's going on and change the underlying code until that test _does_ pass

this is called unit testing and I recommend writing a test like this whenever you're like "I am like 90% sure I'm doing everything right and I am not getting the results I expect"

debugging in a systematic way is extremely effective, and this is part of that

(also it helps me figure out what's going on because I have a concrete example to work with)

for more advanced setups (GitHub actually makes it pretty easy), you can set these tests to automatically run on a branch whenever a Pull Request is open on it, and you can see right on the GitHub page which tests pass and fail

that's automated testing, and most tech companies make sure all tests pass for any branch/PR that gets merged to `main`, which does a lot to prevent bugs

Reply:
So I see what you're saying and yes the test_hash_stuff function passes but like if I print the hex string I return from hash_test() then every time I run the function through the cli I get a different string. So my file names will still be different?

Reply:
so the output of `hash_test()` depends on the inputs
if you're typing things differently every time (space, no space, 5 digit zip vs 9 digit zip, capitalization, etc) it'll be different

I recommend writing everything _as if_ `hash_test()` functions correctly, even if it doesn't

and then once everything is in place, we can see what's actually going on

because the test passes, I can't help you figure out what is different from your expectation; writing tests is a way to codify the expected behavior of the code, and to diagnose when the actual behavior differs from the expectation

without a test, I don't know whether your expectation is wrong, or your expectation is correct but you've written a bug somewhere, or something else